sw/host/rom_ext_image_tools/signer/src/main.rs

```diff
@@ -9,91 +9,188 @@
-    image.update_static_fields(&config);
-
-    // Convert ASN.1 DER private key into Mundane RsaPrivKey.
-    let private_key = RsaPrivKey::parse_from_der(&private_key_der)
-        .expect("Failed to parse private key!");
-
-    let mut exponent = private_key.public_exponent_be();
-    // Encode public key components in little-endian byte order.
-    exponent.reverse();
-    image.update_exponent_field(&exponent);
-
-    let mut modulus = private_key.public_modulus_be();
-    // Encode public key components in little-endian byte order.
-    modulus.reverse();
-    image.update_modulus_field(&modulus);
-
-    // Produce the signature from concatenated system_state_value,
-    // device_usage_value and the portion of the "signed" portion of the image.
-    image.update_timestamp_field();
-    let image_sign_data = image.data_to_sign();
-    let device_usage_value =
-        &device_usage_value(&config.input_files.usage_constraints_path);
-    let system_state_value =
-        &system_state_value(&config.input_files.system_state_value_path);
+struct Args {
+    input: PathBuf,
+    priv_key: PathBuf,
+    output: PathBuf,
+}
 
-    let mut message_to_sign = Vec::<u8>::new();
-    message_to_sign.extend_from_slice(system_state_value);
-    message_to_sign.extend_from_slice(device_usage_value);
-    message_to_sign.extend_from_slice(image_sign_data);
+impl Args {
+    pub fn new(args: env::Args) -> Result<Args> {
+        let args = args.skip(1).collect::<Vec<_>>();
+        match args.as_slice() {
+            [input, priv_key, output] => Ok(Args {
+                    input: PathBuf::from(input),
+                    priv_key: PathBuf::from(priv_key),
+                    output: PathBuf::from(output),
+                }),
+            args => bail!("Expected exactly 3 positional arguments: input, private key, and output, got: {}.", args.len()),
+        }
+    }
+}
 
-    let signature = RsaSignature::<B3072, RsaPkcs1v15, Sha256>::sign(
-        &private_key,
-        &message_to_sign,
-    )
-    .expect("Failed to sign!");
+/// Parses an unsigned big-endian hex string into a little-endian byte vector.
+fn parse_hex_str(hex_str: &str) -> Result<Vec<u8>> {
+    ensure!(
+        hex_str.starts_with("0x") && hex_str.len() >= 3,
+        "Invalid hex string: {}",
+        hex_str
+    );
+
+    let mut bytes = Vec::new();
+    hex_str
+        .as_bytes()
+        .iter()
+        .skip(2)
+        .rev()
+        .map(|digit| match digit {
+            b'0'..=b'9' => Ok(digit - b'0'),
+            b'a'..=b'f' => Ok(digit - b'a' + 10),
+            b'A'..=b'F' => Ok(digit - b'A' + 10),
+            _ => bail!("Invalid hex digit: {} ({}).", *digit as char, digit),
+        })
+        .enumerate()
+        .try_for_each(|(i, digit)| -> Result<()> {
+            match i % 2 {
+                0 => bytes.push(digit?),
+                _ => *bytes.last_mut().unwrap() |= digit? << 4,
+            };
+            Ok(())
+        })?;
```

Comment:
I think I might write something like this:

```rust
let bytes = hex_str.as_bytes()[2..];
ensure(bytes.len() % 2 == 0);
let mut out = Vec::with_capacity(bytes.len() / 2);
bytes.chunks(2).map(|digits| {
  hex_digit(digits[0])? << 4 | hex_digit(digits[1])?
}).rev().try_for_each(|byte| { out.push(byte?); Ok(()) });
```

Reply:
Ah, I like `chunks` but ended up using `from_str_radix` for this. PTAL.